tests/scripts/thread-cert/test_netdata_publisher.py

```diff
@@ -0,0 +1,456 @@
+#!/usr/bin/env python3
+#
+#  Copyright (c) 2021, The OpenThread Authors.
+#  All rights reserved.
+#
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions are met:
+#  1. Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+#  2. Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#  3. Neither the name of the copyright holder nor the
+#     names of its contributors may be used to endorse or promote products
+#     derived from this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+import ipaddress
+import unittest
+
+import command
+import thread_cert
+
+# Test description:
+#   This test verifies network data publisher behavior with DNS/SRP service entries and on-mesh prefix and external
+#   route entries.
+#
+# Topology:
+#
+#   1 leader, 5 routers and 5 end-devices all connected.
+#
+
+LEADER = 1
+ROUTER1 = 2
+ROUTER2 = 3
+ROUTER3 = 4
+ROUTER4 = 5
+ROUTER5 = 6
+END_DEV1 = 7
+END_DEV2 = 8
+END_DEV3 = 9
+END_DEV4 = 10
+END_DEV5 = 11
+
+WAIT_TIME = 55
+
+ON_MESH_PREFIX = 'fd00:1234::/64'
+ON_MESH_FLAGS = 'paso'
+
+EXTERNAL_ROUTE = 'fd00:abce::/64'
+EXTERNAL_FLAGS = 's'
+
+ANYCAST_SEQ_NUM = '4'
+
+DNSSRP_ADDRESS = 'fd00::cdef'
+DNSSRP_PORT = '8765'
+
+# The desired number of entries (based on related config).
+DESIRED_NUM_DNSSRP_ANYCAST = 8
+DESIRED_NUM_DNSSRP_UNCIAST = 2
+DESIRED_NUM_ON_MESH_PREFIX = 3
+DESIRED_NUM_EXTERNAL_ROUTE = 10
+
+THREAD_ENTERPRISE_NUMBER = 44970
+ANYCAST_SERVICE_NUM = 0x5c
+UNICAST_SERVICE_NUM = 0x5d
+
+
+def parse_hex_string(hexstr: str) -> bytes:
+    return bytes(int(hexstr[i:i + 2], 16) for i in range(0, len(hexstr), 2))
```

Comment:
We can use [bytes.from_string()](https://docs.python.org/3/library/stdtypes.html#bytes.fromhex).

Reply:
Changed in the new push (actually removed the function and directly used the `bytes.fromhex()`.